{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from scipy import stats\n",
    "\n",
    "import statsmodels.formula.api as smf\n",
    "\n",
    "from matplotlib import pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Import data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This data comes from this paper: https://www.nature.com/articles/s41467-019-13483-w\n",
    "\n",
    "It's a big excel file with expression and growth rate data in different sheets and this code will use a few of these sheets."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.ExcelFile(r'../Data/raw_data/41467_2019_13483_MOESM4_ESM.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basic metadata clean up and data subsetting\n",
    "\n",
    "After some manual inspection, looking at distributions, etc. we decided to exclude samples where:\n",
    "\n",
    "1. The growth rate data was unknown\n",
    "\n",
    "\n",
    "2. The growth rate data was reported as zero. This one might seem strange but it's is a little unclear in general if that's possible/true to have zero growth rate. It is possible that these were stationary phase cultures but equally likely from my stand-point that these are errors in the table.\n",
    "\n",
    "\n",
    "3. Really poor alignment (perhaps indicating some overall contamination)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(278, 26)\n",
      "(195, 26)\n",
      "(179, 26)\n",
      "(173, 26)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Study</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Condition ID</th>\n",
       "      <th>Replicate #</th>\n",
       "      <th>Strain Description</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Base Media</th>\n",
       "      <th>Carbon Source (g/L)</th>\n",
       "      <th>Nitrogen Source (g/L)</th>\n",
       "      <th>...</th>\n",
       "      <th>Culture Type</th>\n",
       "      <th>Growth Rate (1/hr)</th>\n",
       "      <th>Evolved Sample</th>\n",
       "      <th>Isolate Type</th>\n",
       "      <th>Sequencing Machine</th>\n",
       "      <th>Additional Details</th>\n",
       "      <th>Biological Replicates</th>\n",
       "      <th>Alignment</th>\n",
       "      <th>DOI</th>\n",
       "      <th>GEO</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fur__wt_fe__1</td>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>wt_fe</td>\n",
       "      <td>1</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>...</td>\n",
       "      <td>Batch</td>\n",
       "      <td>1.060606</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>93.35</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>fur__wt_fe__2</td>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>wt_fe</td>\n",
       "      <td>2</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>...</td>\n",
       "      <td>Batch</td>\n",
       "      <td>1.060606</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>92.38</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>fur__delfur_fe2__1</td>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>delfur_fe2</td>\n",
       "      <td>1</td>\n",
       "      <td>Escherichia coli K-12 MG1655 del_fur</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>...</td>\n",
       "      <td>Batch</td>\n",
       "      <td>0.619469</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>92.80</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>fur__delfur_fe2__2</td>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>delfur_fe2</td>\n",
       "      <td>2</td>\n",
       "      <td>Escherichia coli K-12 MG1655 del_fur</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>...</td>\n",
       "      <td>Batch</td>\n",
       "      <td>0.619469</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>93.24</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>omics__bw_ac__1</td>\n",
       "      <td>Omics</td>\n",
       "      <td>omics</td>\n",
       "      <td>bw_ac</td>\n",
       "      <td>1</td>\n",
       "      <td>Escherichia coli BW25113</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>M9</td>\n",
       "      <td>acetate(3.5)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>...</td>\n",
       "      <td>Batch</td>\n",
       "      <td>0.203000</td>\n",
       "      <td>No</td>\n",
       "      <td>NaN</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2</td>\n",
       "      <td>97.80</td>\n",
       "      <td>doi.org/10.1038/ncomms13091</td>\n",
       "      <td>GSE59759</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Sample ID  Study Project ID Condition ID  Replicate #  \\\n",
       "4        fur__wt_fe__1    Fur        fur        wt_fe            1   \n",
       "5        fur__wt_fe__2    Fur        fur        wt_fe            2   \n",
       "8   fur__delfur_fe2__1    Fur        fur   delfur_fe2            1   \n",
       "9   fur__delfur_fe2__2    Fur        fur   delfur_fe2            2   \n",
       "55     omics__bw_ac__1  Omics      omics        bw_ac            1   \n",
       "\n",
       "                      Strain Description   Strain Base Media  \\\n",
       "4           Escherichia coli K-12 MG1655   MG1655         M9   \n",
       "5           Escherichia coli K-12 MG1655   MG1655         M9   \n",
       "8   Escherichia coli K-12 MG1655 del_fur   MG1655         M9   \n",
       "9   Escherichia coli K-12 MG1655 del_fur   MG1655         M9   \n",
       "55              Escherichia coli BW25113  BW25113         M9   \n",
       "\n",
       "   Carbon Source (g/L) Nitrogen Source (g/L)  ... Culture Type  \\\n",
       "4           glucose(2)              NH4Cl(1)  ...        Batch   \n",
       "5           glucose(2)              NH4Cl(1)  ...        Batch   \n",
       "8           glucose(2)              NH4Cl(1)  ...        Batch   \n",
       "9           glucose(2)              NH4Cl(1)  ...        Batch   \n",
       "55        acetate(3.5)              NH4Cl(1)  ...        Batch   \n",
       "\n",
       "   Growth Rate (1/hr) Evolved Sample  Isolate Type  Sequencing Machine  \\\n",
       "4            1.060606             No           NaN               MiSeq   \n",
       "5            1.060606             No           NaN               MiSeq   \n",
       "8            0.619469             No           NaN               MiSeq   \n",
       "9            0.619469             No           NaN               MiSeq   \n",
       "55           0.203000             No           NaN               MiSeq   \n",
       "\n",
       "   Additional Details Biological Replicates  Alignment  \\\n",
       "4                 NaN                     2      93.35   \n",
       "5                 NaN                     2      92.38   \n",
       "8                 NaN                     2      92.80   \n",
       "9                 NaN                     2      93.24   \n",
       "55                NaN                     2      97.80   \n",
       "\n",
       "                            DOI       GEO  \n",
       "4    doi.org/10.1038/ncomms5910  GSE54900  \n",
       "5    doi.org/10.1038/ncomms5910  GSE54900  \n",
       "8    doi.org/10.1038/ncomms5910  GSE54900  \n",
       "9    doi.org/10.1038/ncomms5910  GSE54900  \n",
       "55  doi.org/10.1038/ncomms13091  GSE59759  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df = df.parse('Metadata') ###This grabs the sheet that contains information about the samples\n",
    "print(meta_df.shape)\n",
    "meta_df = meta_df[meta_df['Growth Rate (1/hr)'].isnull() == False]\n",
    "print(meta_df.shape)\n",
    "meta_df = meta_df[meta_df['Growth Rate (1/hr)'] > 0.0]\n",
    "print(meta_df.shape)\n",
    "meta_df = meta_df[meta_df['Alignment'] > 80]\n",
    "print(meta_df.shape)\n",
    "meta_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Read in the expression data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3923, 278)\n",
      "(3923, 173)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fur__wt_fe__1</th>\n",
       "      <th>fur__wt_fe__2</th>\n",
       "      <th>fur__delfur_fe2__1</th>\n",
       "      <th>fur__delfur_fe2__2</th>\n",
       "      <th>omics__bw_ac__1</th>\n",
       "      <th>omics__bw_ac__2</th>\n",
       "      <th>omics__bw_fum__1</th>\n",
       "      <th>omics__bw_fum__2</th>\n",
       "      <th>omics__bw_glc__1</th>\n",
       "      <th>omics__bw_glc__2</th>\n",
       "      <th>...</th>\n",
       "      <th>efeU__menFentC_ale29__1</th>\n",
       "      <th>efeU__menFentC_ale29__2</th>\n",
       "      <th>efeU__menFentC_ale30__1</th>\n",
       "      <th>efeU__menFentC_ale30__2</th>\n",
       "      <th>efeU__menFentCubiC_ale36__1</th>\n",
       "      <th>efeU__menFentCubiC_ale36__2</th>\n",
       "      <th>efeU__menFentCubiC_ale37__1</th>\n",
       "      <th>efeU__menFentCubiC_ale37__2</th>\n",
       "      <th>efeU__menFentCubiC_ale38__1</th>\n",
       "      <th>efeU__menFentCubiC_ale38__2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log-TPM</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>11.064810</td>\n",
       "      <td>10.779071</td>\n",
       "      <td>11.229767</td>\n",
       "      <td>11.214065</td>\n",
       "      <td>9.257348</td>\n",
       "      <td>9.182322</td>\n",
       "      <td>9.709213</td>\n",
       "      <td>9.672126</td>\n",
       "      <td>10.208587</td>\n",
       "      <td>10.218351</td>\n",
       "      <td>...</td>\n",
       "      <td>10.271327</td>\n",
       "      <td>10.276565</td>\n",
       "      <td>11.148538</td>\n",
       "      <td>11.170578</td>\n",
       "      <td>11.676604</td>\n",
       "      <td>11.726097</td>\n",
       "      <td>11.881529</td>\n",
       "      <td>11.923237</td>\n",
       "      <td>11.496410</td>\n",
       "      <td>11.552762</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>10.776984</td>\n",
       "      <td>10.597810</td>\n",
       "      <td>10.897938</td>\n",
       "      <td>10.861157</td>\n",
       "      <td>8.983408</td>\n",
       "      <td>8.943151</td>\n",
       "      <td>9.436004</td>\n",
       "      <td>9.394573</td>\n",
       "      <td>9.609637</td>\n",
       "      <td>9.677931</td>\n",
       "      <td>...</td>\n",
       "      <td>10.160291</td>\n",
       "      <td>10.116861</td>\n",
       "      <td>10.314322</td>\n",
       "      <td>10.392251</td>\n",
       "      <td>10.916426</td>\n",
       "      <td>10.909277</td>\n",
       "      <td>11.023924</td>\n",
       "      <td>11.038426</td>\n",
       "      <td>10.624301</td>\n",
       "      <td>10.764195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>10.394971</td>\n",
       "      <td>10.113950</td>\n",
       "      <td>10.185151</td>\n",
       "      <td>10.164655</td>\n",
       "      <td>8.761690</td>\n",
       "      <td>8.779920</td>\n",
       "      <td>9.532673</td>\n",
       "      <td>9.534370</td>\n",
       "      <td>9.883558</td>\n",
       "      <td>9.870356</td>\n",
       "      <td>...</td>\n",
       "      <td>10.475069</td>\n",
       "      <td>10.434352</td>\n",
       "      <td>10.679541</td>\n",
       "      <td>10.723953</td>\n",
       "      <td>11.143310</td>\n",
       "      <td>11.112721</td>\n",
       "      <td>11.184795</td>\n",
       "      <td>11.241845</td>\n",
       "      <td>10.953206</td>\n",
       "      <td>11.001006</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>6.716069</td>\n",
       "      <td>6.410864</td>\n",
       "      <td>6.527653</td>\n",
       "      <td>6.136168</td>\n",
       "      <td>4.474204</td>\n",
       "      <td>4.720490</td>\n",
       "      <td>5.782102</td>\n",
       "      <td>5.326669</td>\n",
       "      <td>5.846675</td>\n",
       "      <td>5.972022</td>\n",
       "      <td>...</td>\n",
       "      <td>5.979079</td>\n",
       "      <td>5.705586</td>\n",
       "      <td>6.306120</td>\n",
       "      <td>6.291340</td>\n",
       "      <td>5.058537</td>\n",
       "      <td>4.835550</td>\n",
       "      <td>5.448097</td>\n",
       "      <td>5.757951</td>\n",
       "      <td>5.873964</td>\n",
       "      <td>5.808618</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>6.761813</td>\n",
       "      <td>6.816532</td>\n",
       "      <td>6.862147</td>\n",
       "      <td>6.817480</td>\n",
       "      <td>6.536457</td>\n",
       "      <td>6.439917</td>\n",
       "      <td>6.408731</td>\n",
       "      <td>6.276017</td>\n",
       "      <td>6.910200</td>\n",
       "      <td>6.843384</td>\n",
       "      <td>...</td>\n",
       "      <td>8.371287</td>\n",
       "      <td>8.322390</td>\n",
       "      <td>8.137515</td>\n",
       "      <td>8.071837</td>\n",
       "      <td>7.354131</td>\n",
       "      <td>7.365536</td>\n",
       "      <td>7.328101</td>\n",
       "      <td>7.314761</td>\n",
       "      <td>8.057750</td>\n",
       "      <td>8.105213</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 173 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         fur__wt_fe__1  fur__wt_fe__2  fur__delfur_fe2__1  fur__delfur_fe2__2  \\\n",
       "log-TPM                                                                         \n",
       "b0002        11.064810      10.779071           11.229767           11.214065   \n",
       "b0003        10.776984      10.597810           10.897938           10.861157   \n",
       "b0004        10.394971      10.113950           10.185151           10.164655   \n",
       "b0005         6.716069       6.410864            6.527653            6.136168   \n",
       "b0006         6.761813       6.816532            6.862147            6.817480   \n",
       "\n",
       "         omics__bw_ac__1  omics__bw_ac__2  omics__bw_fum__1  omics__bw_fum__2  \\\n",
       "log-TPM                                                                         \n",
       "b0002           9.257348         9.182322          9.709213          9.672126   \n",
       "b0003           8.983408         8.943151          9.436004          9.394573   \n",
       "b0004           8.761690         8.779920          9.532673          9.534370   \n",
       "b0005           4.474204         4.720490          5.782102          5.326669   \n",
       "b0006           6.536457         6.439917          6.408731          6.276017   \n",
       "\n",
       "         omics__bw_glc__1  omics__bw_glc__2  ...  efeU__menFentC_ale29__1  \\\n",
       "log-TPM                                      ...                            \n",
       "b0002           10.208587         10.218351  ...                10.271327   \n",
       "b0003            9.609637          9.677931  ...                10.160291   \n",
       "b0004            9.883558          9.870356  ...                10.475069   \n",
       "b0005            5.846675          5.972022  ...                 5.979079   \n",
       "b0006            6.910200          6.843384  ...                 8.371287   \n",
       "\n",
       "         efeU__menFentC_ale29__2  efeU__menFentC_ale30__1  \\\n",
       "log-TPM                                                     \n",
       "b0002                  10.276565                11.148538   \n",
       "b0003                  10.116861                10.314322   \n",
       "b0004                  10.434352                10.679541   \n",
       "b0005                   5.705586                 6.306120   \n",
       "b0006                   8.322390                 8.137515   \n",
       "\n",
       "         efeU__menFentC_ale30__2  efeU__menFentCubiC_ale36__1  \\\n",
       "log-TPM                                                         \n",
       "b0002                  11.170578                    11.676604   \n",
       "b0003                  10.392251                    10.916426   \n",
       "b0004                  10.723953                    11.143310   \n",
       "b0005                   6.291340                     5.058537   \n",
       "b0006                   8.071837                     7.354131   \n",
       "\n",
       "         efeU__menFentCubiC_ale36__2  efeU__menFentCubiC_ale37__1  \\\n",
       "log-TPM                                                             \n",
       "b0002                      11.726097                    11.881529   \n",
       "b0003                      10.909277                    11.023924   \n",
       "b0004                      11.112721                    11.184795   \n",
       "b0005                       4.835550                     5.448097   \n",
       "b0006                       7.365536                     7.328101   \n",
       "\n",
       "         efeU__menFentCubiC_ale37__2  efeU__menFentCubiC_ale38__1  \\\n",
       "log-TPM                                                             \n",
       "b0002                      11.923237                    11.496410   \n",
       "b0003                      11.038426                    10.624301   \n",
       "b0004                      11.241845                    10.953206   \n",
       "b0005                       5.757951                     5.873964   \n",
       "b0006                       7.314761                     8.057750   \n",
       "\n",
       "         efeU__menFentCubiC_ale38__2  \n",
       "log-TPM                               \n",
       "b0002                      11.552762  \n",
       "b0003                      10.764195  \n",
       "b0004                      11.001006  \n",
       "b0005                       5.808618  \n",
       "b0006                       8.105213  \n",
       "\n",
       "[5 rows x 173 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_df = df.parse('Expression Data', index_col='log-TPM')\n",
    "print(exp_df.shape)\n",
    "exp_df = exp_df[meta_df['Sample ID']]###Only grab the columns corresponding to the samples identified above\n",
    "print(exp_df.shape)\n",
    "assert list(meta_df['Sample ID']) == list(exp_df.columns) ###Check our work\n",
    "exp_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Manual inspection found some weird and highly implausible/impossible duplicate column/s, so we'll make a note of that here and deal with it later**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pal__lyx_ale2_f6__1</th>\n",
       "      <th>pal__lyx_ale2__1</th>\n",
       "      <th>pal__lyx_ale2__2</th>\n",
       "      <th>pal__lyx_ale4__1</th>\n",
       "      <th>pal__lyx_ale4__2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log-TPM</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>9.627287</td>\n",
       "      <td>10.130315</td>\n",
       "      <td>10.155462</td>\n",
       "      <td>10.130315</td>\n",
       "      <td>10.155462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>9.250534</td>\n",
       "      <td>9.708944</td>\n",
       "      <td>9.831138</td>\n",
       "      <td>9.708944</td>\n",
       "      <td>9.831138</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>9.203814</td>\n",
       "      <td>10.049444</td>\n",
       "      <td>10.190627</td>\n",
       "      <td>10.049444</td>\n",
       "      <td>10.190627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>4.807384</td>\n",
       "      <td>5.772047</td>\n",
       "      <td>5.933463</td>\n",
       "      <td>5.772047</td>\n",
       "      <td>5.933463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>6.398236</td>\n",
       "      <td>6.435048</td>\n",
       "      <td>6.220552</td>\n",
       "      <td>6.435048</td>\n",
       "      <td>6.220552</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         pal__lyx_ale2_f6__1  pal__lyx_ale2__1  pal__lyx_ale2__2  \\\n",
       "log-TPM                                                            \n",
       "b0002               9.627287         10.130315         10.155462   \n",
       "b0003               9.250534          9.708944          9.831138   \n",
       "b0004               9.203814         10.049444         10.190627   \n",
       "b0005               4.807384          5.772047          5.933463   \n",
       "b0006               6.398236          6.435048          6.220552   \n",
       "\n",
       "         pal__lyx_ale4__1  pal__lyx_ale4__2  \n",
       "log-TPM                                      \n",
       "b0002           10.130315         10.155462  \n",
       "b0003            9.708944          9.831138  \n",
       "b0004           10.049444         10.190627  \n",
       "b0005            5.772047          5.933463  \n",
       "b0006            6.435048          6.220552  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exp_df[[col for col in exp_df.columns if 'pal__lyx_ale' in col]].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Averaging gene expression values across replicates to clean up/simplify the data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## First getting gene expression averages between replicates and creating a new `dataframe` to hold this information\n",
    "\n",
    "Thus just exploits the fact that replicates are denoted by \"__x\" in the sample names so we identify these and average them when possible. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3923, 105)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ytf__delyafC</th>\n",
       "      <th>ytf__delyddM</th>\n",
       "      <th>pgi__glu4_delpgi</th>\n",
       "      <th>pgi__pgi_ale7</th>\n",
       "      <th>fur__delfur_fe2</th>\n",
       "      <th>pgi__pgi_ale8</th>\n",
       "      <th>pal__lyx_ale4</th>\n",
       "      <th>ica__leu_glcr</th>\n",
       "      <th>rpoB__rpoBE546V_glc</th>\n",
       "      <th>efeU__menFentC_ale39</th>\n",
       "      <th>...</th>\n",
       "      <th>42c__42c_ale5</th>\n",
       "      <th>pgi__pgi_ale6</th>\n",
       "      <th>ytf__delyheO</th>\n",
       "      <th>pal__arab_ale14_f9</th>\n",
       "      <th>cra_crp__wt_fru</th>\n",
       "      <th>glu__glu_ale5</th>\n",
       "      <th>ssw__glc_xyl_glc</th>\n",
       "      <th>ica__thm_gal</th>\n",
       "      <th>pal__arab_ale16</th>\n",
       "      <th>ytf__wt_ph8</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log-TPM</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>9.312825</td>\n",
       "      <td>11.112830</td>\n",
       "      <td>9.846088</td>\n",
       "      <td>10.862762</td>\n",
       "      <td>11.221916</td>\n",
       "      <td>10.482452</td>\n",
       "      <td>10.142889</td>\n",
       "      <td>12.361644</td>\n",
       "      <td>10.927354</td>\n",
       "      <td>10.344302</td>\n",
       "      <td>...</td>\n",
       "      <td>10.627727</td>\n",
       "      <td>9.859055</td>\n",
       "      <td>9.401156</td>\n",
       "      <td>10.227775</td>\n",
       "      <td>11.198040</td>\n",
       "      <td>11.139383</td>\n",
       "      <td>10.079777</td>\n",
       "      <td>11.090470</td>\n",
       "      <td>8.970781</td>\n",
       "      <td>11.135983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>8.680662</td>\n",
       "      <td>10.412806</td>\n",
       "      <td>9.253330</td>\n",
       "      <td>10.194297</td>\n",
       "      <td>10.879548</td>\n",
       "      <td>9.744255</td>\n",
       "      <td>9.770041</td>\n",
       "      <td>11.403576</td>\n",
       "      <td>10.293011</td>\n",
       "      <td>10.265772</td>\n",
       "      <td>...</td>\n",
       "      <td>10.015906</td>\n",
       "      <td>9.214803</td>\n",
       "      <td>8.933743</td>\n",
       "      <td>9.808444</td>\n",
       "      <td>10.182029</td>\n",
       "      <td>10.650042</td>\n",
       "      <td>9.672696</td>\n",
       "      <td>10.402663</td>\n",
       "      <td>8.564640</td>\n",
       "      <td>10.507847</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>8.735598</td>\n",
       "      <td>10.728984</td>\n",
       "      <td>9.273931</td>\n",
       "      <td>10.104900</td>\n",
       "      <td>10.174903</td>\n",
       "      <td>9.761211</td>\n",
       "      <td>10.120036</td>\n",
       "      <td>11.844689</td>\n",
       "      <td>10.246893</td>\n",
       "      <td>10.712971</td>\n",
       "      <td>...</td>\n",
       "      <td>10.239207</td>\n",
       "      <td>9.128370</td>\n",
       "      <td>9.164724</td>\n",
       "      <td>9.854127</td>\n",
       "      <td>10.037740</td>\n",
       "      <td>10.828312</td>\n",
       "      <td>9.735267</td>\n",
       "      <td>10.633509</td>\n",
       "      <td>8.875950</td>\n",
       "      <td>10.650490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>4.190217</td>\n",
       "      <td>5.831692</td>\n",
       "      <td>4.381576</td>\n",
       "      <td>6.461725</td>\n",
       "      <td>6.331911</td>\n",
       "      <td>6.083744</td>\n",
       "      <td>5.852755</td>\n",
       "      <td>6.207579</td>\n",
       "      <td>7.551687</td>\n",
       "      <td>6.245881</td>\n",
       "      <td>...</td>\n",
       "      <td>5.799398</td>\n",
       "      <td>6.575517</td>\n",
       "      <td>4.215120</td>\n",
       "      <td>6.178483</td>\n",
       "      <td>6.058802</td>\n",
       "      <td>6.753118</td>\n",
       "      <td>6.280117</td>\n",
       "      <td>4.686068</td>\n",
       "      <td>5.094244</td>\n",
       "      <td>5.651241</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>6.769807</td>\n",
       "      <td>7.205476</td>\n",
       "      <td>7.260331</td>\n",
       "      <td>6.924540</td>\n",
       "      <td>6.839814</td>\n",
       "      <td>6.296399</td>\n",
       "      <td>6.327800</td>\n",
       "      <td>6.780668</td>\n",
       "      <td>6.882767</td>\n",
       "      <td>7.161943</td>\n",
       "      <td>...</td>\n",
       "      <td>6.945713</td>\n",
       "      <td>6.865126</td>\n",
       "      <td>6.748235</td>\n",
       "      <td>6.589387</td>\n",
       "      <td>6.555468</td>\n",
       "      <td>6.723685</td>\n",
       "      <td>6.895570</td>\n",
       "      <td>6.663905</td>\n",
       "      <td>6.655641</td>\n",
       "      <td>7.034334</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 105 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         ytf__delyafC  ytf__delyddM  pgi__glu4_delpgi  pgi__pgi_ale7  \\\n",
       "log-TPM                                                                \n",
       "b0002        9.312825     11.112830          9.846088      10.862762   \n",
       "b0003        8.680662     10.412806          9.253330      10.194297   \n",
       "b0004        8.735598     10.728984          9.273931      10.104900   \n",
       "b0005        4.190217      5.831692          4.381576       6.461725   \n",
       "b0006        6.769807      7.205476          7.260331       6.924540   \n",
       "\n",
       "         fur__delfur_fe2  pgi__pgi_ale8  pal__lyx_ale4  ica__leu_glcr  \\\n",
       "log-TPM                                                                 \n",
       "b0002          11.221916      10.482452      10.142889      12.361644   \n",
       "b0003          10.879548       9.744255       9.770041      11.403576   \n",
       "b0004          10.174903       9.761211      10.120036      11.844689   \n",
       "b0005           6.331911       6.083744       5.852755       6.207579   \n",
       "b0006           6.839814       6.296399       6.327800       6.780668   \n",
       "\n",
       "         rpoB__rpoBE546V_glc  efeU__menFentC_ale39  ...  42c__42c_ale5  \\\n",
       "log-TPM                                             ...                  \n",
       "b0002              10.927354             10.344302  ...      10.627727   \n",
       "b0003              10.293011             10.265772  ...      10.015906   \n",
       "b0004              10.246893             10.712971  ...      10.239207   \n",
       "b0005               7.551687              6.245881  ...       5.799398   \n",
       "b0006               6.882767              7.161943  ...       6.945713   \n",
       "\n",
       "         pgi__pgi_ale6  ytf__delyheO  pal__arab_ale14_f9  cra_crp__wt_fru  \\\n",
       "log-TPM                                                                     \n",
       "b0002         9.859055      9.401156           10.227775        11.198040   \n",
       "b0003         9.214803      8.933743            9.808444        10.182029   \n",
       "b0004         9.128370      9.164724            9.854127        10.037740   \n",
       "b0005         6.575517      4.215120            6.178483         6.058802   \n",
       "b0006         6.865126      6.748235            6.589387         6.555468   \n",
       "\n",
       "         glu__glu_ale5  ssw__glc_xyl_glc  ica__thm_gal  pal__arab_ale16  \\\n",
       "log-TPM                                                                   \n",
       "b0002        11.139383         10.079777     11.090470         8.970781   \n",
       "b0003        10.650042          9.672696     10.402663         8.564640   \n",
       "b0004        10.828312          9.735267     10.633509         8.875950   \n",
       "b0005         6.753118          6.280117      4.686068         5.094244   \n",
       "b0006         6.723685          6.895570      6.663905         6.655641   \n",
       "\n",
       "         ytf__wt_ph8  \n",
       "log-TPM               \n",
       "b0002      11.135983  \n",
       "b0003      10.507847  \n",
       "b0004      10.650490  \n",
       "b0005       5.651241  \n",
       "b0006       7.034334  \n",
       "\n",
       "[5 rows x 105 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_cols = set(exp_df.columns.str[:-3])\n",
    "new_exp_df = pd.DataFrame()\n",
    "\n",
    "for i in unique_cols:\n",
    "    new_exp_df[i] = exp_df[[col for col in exp_df.columns if col[:-3]==i]].mean(axis=1)\n",
    "print(new_exp_df.shape)\n",
    "new_exp_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Double checking the work**\n",
    "\n",
    "Just making sure somethings add up here by taking an example column/condition and looking at the replicate values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ytf__delyafC__1</th>\n",
       "      <th>ytf__delyafC__2</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log-TPM</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>9.378698</td>\n",
       "      <td>9.246952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>8.727473</td>\n",
       "      <td>8.633851</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>8.745293</td>\n",
       "      <td>8.725903</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>4.154798</td>\n",
       "      <td>4.225636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>6.826286</td>\n",
       "      <td>6.713329</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ytf__delyafC__1  ytf__delyafC__2\n",
       "log-TPM                                  \n",
       "b0002           9.378698         9.246952\n",
       "b0003           8.727473         8.633851\n",
       "b0004           8.745293         8.725903\n",
       "b0005           4.154798         4.225636\n",
       "b0006           6.826286         6.713329"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "example_col = list(unique_cols)[0]\n",
    "\n",
    "exp_df[[col for col in exp_df.columns if col[:-3]==example_col]].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And their average"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ytf__delyafC</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>log-TPM</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0002</th>\n",
       "      <td>9.312825</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0003</th>\n",
       "      <td>8.680662</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0004</th>\n",
       "      <td>8.735598</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0005</th>\n",
       "      <td>4.190217</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0006</th>\n",
       "      <td>6.769807</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         ytf__delyafC\n",
       "log-TPM              \n",
       "b0002        9.312825\n",
       "b0003        8.680662\n",
       "b0004        8.735598\n",
       "b0005        4.190217\n",
       "b0006        6.769807"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_exp_df[[example_col]].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Dealing with the weird duplicate column/s\n",
    "\n",
    "Time to kill any completely identical columns, these must be bugs on the data end and even though I could in theory keep one, their growth rate value is unclear (since it differs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Get an all-to-all correlation matrix between gene expression values\n",
    "temp_corr = new_exp_df.corr(method='spearman')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(105, 105)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ytf__delyafC</th>\n",
       "      <th>ytf__delyddM</th>\n",
       "      <th>pgi__glu4_delpgi</th>\n",
       "      <th>pgi__pgi_ale7</th>\n",
       "      <th>fur__delfur_fe2</th>\n",
       "      <th>pgi__pgi_ale8</th>\n",
       "      <th>pal__lyx_ale4</th>\n",
       "      <th>ica__leu_glcr</th>\n",
       "      <th>rpoB__rpoBE546V_glc</th>\n",
       "      <th>efeU__menFentC_ale39</th>\n",
       "      <th>...</th>\n",
       "      <th>42c__42c_ale5</th>\n",
       "      <th>pgi__pgi_ale6</th>\n",
       "      <th>ytf__delyheO</th>\n",
       "      <th>pal__arab_ale14_f9</th>\n",
       "      <th>cra_crp__wt_fru</th>\n",
       "      <th>glu__glu_ale5</th>\n",
       "      <th>ssw__glc_xyl_glc</th>\n",
       "      <th>ica__thm_gal</th>\n",
       "      <th>pal__arab_ale16</th>\n",
       "      <th>ytf__wt_ph8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ytf__delyafC</th>\n",
       "      <td>NaN</td>\n",
       "      <td>0.917379</td>\n",
       "      <td>0.822837</td>\n",
       "      <td>0.824888</td>\n",
       "      <td>0.856039</td>\n",
       "      <td>0.829152</td>\n",
       "      <td>0.799936</td>\n",
       "      <td>0.791314</td>\n",
       "      <td>0.849476</td>\n",
       "      <td>0.938303</td>\n",
       "      <td>...</td>\n",
       "      <td>0.803504</td>\n",
       "      <td>0.815242</td>\n",
       "      <td>0.990424</td>\n",
       "      <td>0.872424</td>\n",
       "      <td>0.852303</td>\n",
       "      <td>0.812769</td>\n",
       "      <td>0.892314</td>\n",
       "      <td>0.822446</td>\n",
       "      <td>0.806385</td>\n",
       "      <td>0.886383</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ytf__delyddM</th>\n",
       "      <td>0.917379</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.886031</td>\n",
       "      <td>0.904277</td>\n",
       "      <td>0.928504</td>\n",
       "      <td>0.911653</td>\n",
       "      <td>0.871243</td>\n",
       "      <td>0.885556</td>\n",
       "      <td>0.929653</td>\n",
       "      <td>0.907190</td>\n",
       "      <td>...</td>\n",
       "      <td>0.893792</td>\n",
       "      <td>0.898835</td>\n",
       "      <td>0.915217</td>\n",
       "      <td>0.936786</td>\n",
       "      <td>0.923216</td>\n",
       "      <td>0.880051</td>\n",
       "      <td>0.894655</td>\n",
       "      <td>0.884198</td>\n",
       "      <td>0.883831</td>\n",
       "      <td>0.959916</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pgi__glu4_delpgi</th>\n",
       "      <td>0.822837</td>\n",
       "      <td>0.886031</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.943960</td>\n",
       "      <td>0.914664</td>\n",
       "      <td>0.934235</td>\n",
       "      <td>0.893330</td>\n",
       "      <td>0.881568</td>\n",
       "      <td>0.913203</td>\n",
       "      <td>0.849512</td>\n",
       "      <td>...</td>\n",
       "      <td>0.905331</td>\n",
       "      <td>0.934927</td>\n",
       "      <td>0.819924</td>\n",
       "      <td>0.904554</td>\n",
       "      <td>0.905393</td>\n",
       "      <td>0.920590</td>\n",
       "      <td>0.848005</td>\n",
       "      <td>0.890937</td>\n",
       "      <td>0.873158</td>\n",
       "      <td>0.909069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pgi__pgi_ale7</th>\n",
       "      <td>0.824888</td>\n",
       "      <td>0.904277</td>\n",
       "      <td>0.943960</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.916229</td>\n",
       "      <td>0.978946</td>\n",
       "      <td>0.911686</td>\n",
       "      <td>0.880287</td>\n",
       "      <td>0.948027</td>\n",
       "      <td>0.850431</td>\n",
       "      <td>...</td>\n",
       "      <td>0.939316</td>\n",
       "      <td>0.979767</td>\n",
       "      <td>0.823600</td>\n",
       "      <td>0.914642</td>\n",
       "      <td>0.904791</td>\n",
       "      <td>0.940711</td>\n",
       "      <td>0.892932</td>\n",
       "      <td>0.873350</td>\n",
       "      <td>0.889021</td>\n",
       "      <td>0.909327</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fur__delfur_fe2</th>\n",
       "      <td>0.856039</td>\n",
       "      <td>0.928504</td>\n",
       "      <td>0.914664</td>\n",
       "      <td>0.916229</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.915113</td>\n",
       "      <td>0.878752</td>\n",
       "      <td>0.851400</td>\n",
       "      <td>0.948601</td>\n",
       "      <td>0.840142</td>\n",
       "      <td>...</td>\n",
       "      <td>0.899094</td>\n",
       "      <td>0.914963</td>\n",
       "      <td>0.844202</td>\n",
       "      <td>0.935461</td>\n",
       "      <td>0.914061</td>\n",
       "      <td>0.899217</td>\n",
       "      <td>0.866064</td>\n",
       "      <td>0.845009</td>\n",
       "      <td>0.883927</td>\n",
       "      <td>0.965289</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 105 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  ytf__delyafC  ytf__delyddM  pgi__glu4_delpgi  pgi__pgi_ale7  \\\n",
       "ytf__delyafC               NaN      0.917379          0.822837       0.824888   \n",
       "ytf__delyddM          0.917379           NaN          0.886031       0.904277   \n",
       "pgi__glu4_delpgi      0.822837      0.886031               NaN       0.943960   \n",
       "pgi__pgi_ale7         0.824888      0.904277          0.943960            NaN   \n",
       "fur__delfur_fe2       0.856039      0.928504          0.914664       0.916229   \n",
       "\n",
       "                  fur__delfur_fe2  pgi__pgi_ale8  pal__lyx_ale4  \\\n",
       "ytf__delyafC             0.856039       0.829152       0.799936   \n",
       "ytf__delyddM             0.928504       0.911653       0.871243   \n",
       "pgi__glu4_delpgi         0.914664       0.934235       0.893330   \n",
       "pgi__pgi_ale7            0.916229       0.978946       0.911686   \n",
       "fur__delfur_fe2               NaN       0.915113       0.878752   \n",
       "\n",
       "                  ica__leu_glcr  rpoB__rpoBE546V_glc  efeU__menFentC_ale39  \\\n",
       "ytf__delyafC           0.791314             0.849476              0.938303   \n",
       "ytf__delyddM           0.885556             0.929653              0.907190   \n",
       "pgi__glu4_delpgi       0.881568             0.913203              0.849512   \n",
       "pgi__pgi_ale7          0.880287             0.948027              0.850431   \n",
       "fur__delfur_fe2        0.851400             0.948601              0.840142   \n",
       "\n",
       "                  ...  42c__42c_ale5  pgi__pgi_ale6  ytf__delyheO  \\\n",
       "ytf__delyafC      ...       0.803504       0.815242      0.990424   \n",
       "ytf__delyddM      ...       0.893792       0.898835      0.915217   \n",
       "pgi__glu4_delpgi  ...       0.905331       0.934927      0.819924   \n",
       "pgi__pgi_ale7     ...       0.939316       0.979767      0.823600   \n",
       "fur__delfur_fe2   ...       0.899094       0.914963      0.844202   \n",
       "\n",
       "                  pal__arab_ale14_f9  cra_crp__wt_fru  glu__glu_ale5  \\\n",
       "ytf__delyafC                0.872424         0.852303       0.812769   \n",
       "ytf__delyddM                0.936786         0.923216       0.880051   \n",
       "pgi__glu4_delpgi            0.904554         0.905393       0.920590   \n",
       "pgi__pgi_ale7               0.914642         0.904791       0.940711   \n",
       "fur__delfur_fe2             0.935461         0.914061       0.899217   \n",
       "\n",
       "                  ssw__glc_xyl_glc  ica__thm_gal  pal__arab_ale16  ytf__wt_ph8  \n",
       "ytf__delyafC              0.892314      0.822446         0.806385     0.886383  \n",
       "ytf__delyddM              0.894655      0.884198         0.883831     0.959916  \n",
       "pgi__glu4_delpgi          0.848005      0.890937         0.873158     0.909069  \n",
       "pgi__pgi_ale7             0.892932      0.873350         0.889021     0.909327  \n",
       "fur__delfur_fe2           0.866064      0.845009         0.883927     0.965289  \n",
       "\n",
       "[5 rows x 105 columns]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###And make this into a symmetric dataframe\n",
    "temp_df = pd.DataFrame(\n",
    "    np.where(np.equal(*np.indices(temp_corr.shape)), np.nan, temp_corr.values),\n",
    "    temp_corr.index, temp_corr.columns\n",
    ")\n",
    "print(temp_df.shape)\n",
    "temp_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Identify columns that contain a value of \"1.\" since this indicates they have a perfect correlation**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ytf__delyafC</th>\n",
       "      <th>ytf__delyddM</th>\n",
       "      <th>pgi__glu4_delpgi</th>\n",
       "      <th>pgi__pgi_ale7</th>\n",
       "      <th>fur__delfur_fe2</th>\n",
       "      <th>pgi__pgi_ale8</th>\n",
       "      <th>pal__lyx_ale4</th>\n",
       "      <th>ica__leu_glcr</th>\n",
       "      <th>rpoB__rpoBE546V_glc</th>\n",
       "      <th>efeU__menFentC_ale39</th>\n",
       "      <th>...</th>\n",
       "      <th>42c__42c_ale5</th>\n",
       "      <th>pgi__pgi_ale6</th>\n",
       "      <th>ytf__delyheO</th>\n",
       "      <th>pal__arab_ale14_f9</th>\n",
       "      <th>cra_crp__wt_fru</th>\n",
       "      <th>glu__glu_ale5</th>\n",
       "      <th>ssw__glc_xyl_glc</th>\n",
       "      <th>ica__thm_gal</th>\n",
       "      <th>pal__arab_ale16</th>\n",
       "      <th>ytf__wt_ph8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pal__lyx_ale4</th>\n",
       "      <td>0.799936</td>\n",
       "      <td>0.871243</td>\n",
       "      <td>0.89333</td>\n",
       "      <td>0.911686</td>\n",
       "      <td>0.878752</td>\n",
       "      <td>0.905686</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.869209</td>\n",
       "      <td>0.897721</td>\n",
       "      <td>0.829072</td>\n",
       "      <td>...</td>\n",
       "      <td>0.890743</td>\n",
       "      <td>0.907159</td>\n",
       "      <td>0.8044</td>\n",
       "      <td>0.907604</td>\n",
       "      <td>0.90089</td>\n",
       "      <td>0.909602</td>\n",
       "      <td>0.859288</td>\n",
       "      <td>0.883647</td>\n",
       "      <td>0.900231</td>\n",
       "      <td>0.884983</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pal__lyx_ale2</th>\n",
       "      <td>0.799936</td>\n",
       "      <td>0.871243</td>\n",
       "      <td>0.89333</td>\n",
       "      <td>0.911686</td>\n",
       "      <td>0.878752</td>\n",
       "      <td>0.905686</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.869209</td>\n",
       "      <td>0.897721</td>\n",
       "      <td>0.829072</td>\n",
       "      <td>...</td>\n",
       "      <td>0.890743</td>\n",
       "      <td>0.907159</td>\n",
       "      <td>0.8044</td>\n",
       "      <td>0.907604</td>\n",
       "      <td>0.90089</td>\n",
       "      <td>0.909602</td>\n",
       "      <td>0.859288</td>\n",
       "      <td>0.883647</td>\n",
       "      <td>0.900231</td>\n",
       "      <td>0.884983</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 105 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               ytf__delyafC  ytf__delyddM  pgi__glu4_delpgi  pgi__pgi_ale7  \\\n",
       "pal__lyx_ale4      0.799936      0.871243           0.89333       0.911686   \n",
       "pal__lyx_ale2      0.799936      0.871243           0.89333       0.911686   \n",
       "\n",
       "               fur__delfur_fe2  pgi__pgi_ale8  pal__lyx_ale4  ica__leu_glcr  \\\n",
       "pal__lyx_ale4         0.878752       0.905686            NaN       0.869209   \n",
       "pal__lyx_ale2         0.878752       0.905686            1.0       0.869209   \n",
       "\n",
       "               rpoB__rpoBE546V_glc  efeU__menFentC_ale39  ...  42c__42c_ale5  \\\n",
       "pal__lyx_ale4             0.897721              0.829072  ...       0.890743   \n",
       "pal__lyx_ale2             0.897721              0.829072  ...       0.890743   \n",
       "\n",
       "               pgi__pgi_ale6  ytf__delyheO  pal__arab_ale14_f9  \\\n",
       "pal__lyx_ale4       0.907159        0.8044            0.907604   \n",
       "pal__lyx_ale2       0.907159        0.8044            0.907604   \n",
       "\n",
       "               cra_crp__wt_fru  glu__glu_ale5  ssw__glc_xyl_glc  ica__thm_gal  \\\n",
       "pal__lyx_ale4          0.90089       0.909602          0.859288      0.883647   \n",
       "pal__lyx_ale2          0.90089       0.909602          0.859288      0.883647   \n",
       "\n",
       "               pal__arab_ale16  ytf__wt_ph8  \n",
       "pal__lyx_ale4         0.900231     0.884983  \n",
       "pal__lyx_ale2         0.900231     0.884983  \n",
       "\n",
       "[2 rows x 105 columns]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_df[temp_df.values==1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pal__lyx_ale4</th>\n",
       "      <th>pal__lyx_ale2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>pal__lyx_ale4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>pal__lyx_ale2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               pal__lyx_ale4  pal__lyx_ale2\n",
       "pal__lyx_ale4            NaN            1.0\n",
       "pal__lyx_ale2            1.0            NaN"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_df.loc[['pal__lyx_ale4', 'pal__lyx_ale2'], ['pal__lyx_ale4', 'pal__lyx_ale2']]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Get rid of them both since there is obviously an error here somewhere"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3923, 103)\n"
     ]
    }
   ],
   "source": [
    "new_exp_df.drop(['pal__lyx_ale2', 'pal__lyx_ale4'], axis=1, inplace=True)\n",
    "print(new_exp_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Average the growth rates across these replicates in the metadata as well"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(173, 27)\n",
      "(105, 25)\n",
      "(105, 26)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Study</th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Condition ID</th>\n",
       "      <th>Strain Description</th>\n",
       "      <th>Strain</th>\n",
       "      <th>Base Media</th>\n",
       "      <th>Carbon Source (g/L)</th>\n",
       "      <th>Nitrogen Source (g/L)</th>\n",
       "      <th>Electron Acceptor</th>\n",
       "      <th>...</th>\n",
       "      <th>Sequencing Machine</th>\n",
       "      <th>Additional Details</th>\n",
       "      <th>DOI</th>\n",
       "      <th>GEO</th>\n",
       "      <th>Simple_sample_id</th>\n",
       "      <th>Temperature (C)</th>\n",
       "      <th>pH</th>\n",
       "      <th>Growth Rate (1/hr)</th>\n",
       "      <th>Alignment</th>\n",
       "      <th>Replicate #</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>fur__wt_fe__1</td>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>wt_fe</td>\n",
       "      <td>Escherichia coli K-12 MG1655</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "      <td>fur__wt_fe</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.060606</td>\n",
       "      <td>92.865</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>fur__delfur_fe2__1</td>\n",
       "      <td>Fur</td>\n",
       "      <td>fur</td>\n",
       "      <td>delfur_fe2</td>\n",
       "      <td>Escherichia coli K-12 MG1655 del_fur</td>\n",
       "      <td>MG1655</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(2)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doi.org/10.1038/ncomms5910</td>\n",
       "      <td>GSE54900</td>\n",
       "      <td>fur__delfur_fe2</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.619469</td>\n",
       "      <td>93.020</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>omics__bw_ac__1</td>\n",
       "      <td>Omics</td>\n",
       "      <td>omics</td>\n",
       "      <td>bw_ac</td>\n",
       "      <td>Escherichia coli BW25113</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>M9</td>\n",
       "      <td>acetate(3.5)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doi.org/10.1038/ncomms13091</td>\n",
       "      <td>GSE59759</td>\n",
       "      <td>omics__bw_ac</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.203000</td>\n",
       "      <td>97.920</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>omics__bw_fum__1</td>\n",
       "      <td>Omics</td>\n",
       "      <td>omics</td>\n",
       "      <td>bw_fum</td>\n",
       "      <td>Escherichia coli BW25113</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>M9</td>\n",
       "      <td>fumarate(2.8)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doi.org/10.1038/ncomms13091</td>\n",
       "      <td>GSE59759</td>\n",
       "      <td>omics__bw_fum</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.309000</td>\n",
       "      <td>98.130</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>omics__bw_glc__1</td>\n",
       "      <td>Omics</td>\n",
       "      <td>omics</td>\n",
       "      <td>bw_glc</td>\n",
       "      <td>Escherichia coli BW25113</td>\n",
       "      <td>BW25113</td>\n",
       "      <td>M9</td>\n",
       "      <td>glucose(5)</td>\n",
       "      <td>NH4Cl(1)</td>\n",
       "      <td>O2</td>\n",
       "      <td>...</td>\n",
       "      <td>MiSeq</td>\n",
       "      <td>NaN</td>\n",
       "      <td>doi.org/10.1038/ncomms13091</td>\n",
       "      <td>GSE59759</td>\n",
       "      <td>omics__bw_glc</td>\n",
       "      <td>37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.607500</td>\n",
       "      <td>97.875</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Sample ID  Study Project ID Condition ID  \\\n",
       "4        fur__wt_fe__1    Fur        fur        wt_fe   \n",
       "8   fur__delfur_fe2__1    Fur        fur   delfur_fe2   \n",
       "55     omics__bw_ac__1  Omics      omics        bw_ac   \n",
       "57    omics__bw_fum__1  Omics      omics       bw_fum   \n",
       "59    omics__bw_glc__1  Omics      omics       bw_glc   \n",
       "\n",
       "                      Strain Description   Strain Base Media  \\\n",
       "4           Escherichia coli K-12 MG1655   MG1655         M9   \n",
       "8   Escherichia coli K-12 MG1655 del_fur   MG1655         M9   \n",
       "55              Escherichia coli BW25113  BW25113         M9   \n",
       "57              Escherichia coli BW25113  BW25113         M9   \n",
       "59              Escherichia coli BW25113  BW25113         M9   \n",
       "\n",
       "   Carbon Source (g/L) Nitrogen Source (g/L) Electron Acceptor  ...  \\\n",
       "4           glucose(2)              NH4Cl(1)                O2  ...   \n",
       "8           glucose(2)              NH4Cl(1)                O2  ...   \n",
       "55        acetate(3.5)              NH4Cl(1)                O2  ...   \n",
       "57       fumarate(2.8)              NH4Cl(1)                O2  ...   \n",
       "59          glucose(5)              NH4Cl(1)                O2  ...   \n",
       "\n",
       "   Sequencing Machine Additional Details                          DOI  \\\n",
       "4               MiSeq                NaN   doi.org/10.1038/ncomms5910   \n",
       "8               MiSeq                NaN   doi.org/10.1038/ncomms5910   \n",
       "55              MiSeq                NaN  doi.org/10.1038/ncomms13091   \n",
       "57              MiSeq                NaN  doi.org/10.1038/ncomms13091   \n",
       "59              MiSeq                NaN  doi.org/10.1038/ncomms13091   \n",
       "\n",
       "         GEO Simple_sample_id Temperature (C)   pH Growth Rate (1/hr)  \\\n",
       "4   GSE54900       fur__wt_fe              37  7.0           1.060606   \n",
       "8   GSE54900  fur__delfur_fe2              37  7.0           0.619469   \n",
       "55  GSE59759     omics__bw_ac              37  7.0           0.203000   \n",
       "57  GSE59759    omics__bw_fum              37  7.0           0.309000   \n",
       "59  GSE59759    omics__bw_glc              37  7.0           0.607500   \n",
       "\n",
       "   Alignment Replicate #  \n",
       "4     92.865           2  \n",
       "8     93.020           2  \n",
       "55    97.920           2  \n",
       "57    98.130           2  \n",
       "59    97.875           2  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "###Assign a unique id that removes the replicate information\n",
    "meta_df['Simple_sample_id'] = meta_df['Sample ID'].str[:-3]\n",
    "print(meta_df.shape)\n",
    "###Group according to this new id\n",
    "group_cols = ['Simple_sample_id']\n",
    "###For these numeric columns I'll take the mean\n",
    "metric_cols_a = ['Temperature (C)', 'pH', 'Growth Rate (1/hr)', 'Alignment']\n",
    "aggs_a = meta_df.groupby(group_cols)[metric_cols_a].mean()\n",
    "###And for these I'll just grab the count\n",
    "metric_cols_b = ['Replicate #', 'Biological Replicates']\n",
    "aggs_b = meta_df.groupby(group_cols)['Replicate #'].count()\n",
    "\n",
    "###Drop the columns from the original dataframe (we'll add them back in later)\n",
    "meta_df.drop(metric_cols_a, axis=1, inplace=True)\n",
    "meta_df.drop(metric_cols_b, axis=1, inplace=True)\n",
    "###And duplicates\n",
    "meta_df.drop_duplicates(subset=group_cols, keep='first', inplace=True)\n",
    "\n",
    "###Now merge the main dataframe with the grouped ones\n",
    "meta_df = meta_df.merge(right=aggs_a, right_index=True, left_on=group_cols, how='right')\n",
    "print(meta_df.shape)\n",
    "meta_df = meta_df.merge(right=aggs_b, right_index=True, left_on=group_cols, how='right')\n",
    "print(meta_df.shape)\n",
    "meta_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**And get rid of those problematic samples from this dataframe as well**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(105, 26)\n",
      "(104, 26)\n",
      "(103, 26)\n"
     ]
    }
   ],
   "source": [
    "print(meta_df.shape)\n",
    "meta_df = meta_df[meta_df['Sample ID'].str.contains('pal__lyx_ale2__')==False]\n",
    "print(meta_df.shape)\n",
    "meta_df = meta_df[meta_df['Sample ID'].str.contains('pal__lyx_ale4__')==False]\n",
    "print(meta_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Sum should equal the original shape!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2    61\n",
       "1    41\n",
       "6     1\n",
       "Name: Replicate #, dtype: int64"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_df['Replicate #'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Make sure that the columns line up when matching across these two dataframes** "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_exp_df = new_exp_df[meta_df['Simple_sample_id']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert all(new_exp_df.columns == meta_df['Simple_sample_id'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**And add the doubling time just for good measure**\n",
    "\n",
    "Which is just a slight transformation of growth rate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "meta_df['Doubling_time'] = np.log(2)/meta_df['Growth Rate (1/hr)']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construct a third `dataframe` containing gene expression data summary stats\n",
    "\n",
    "Strictly speaking this isn't super necessary but might as well do it now to get it done and over with"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_summary_df = new_exp_df.apply(pd.DataFrame.describe, axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**The % signs seem to cause some problems down the road so lets remove them**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['count', 'mean', 'std', 'min', '25_percentile', '50_percentile', '75_percentile', 'max']\n"
     ]
    }
   ],
   "source": [
    "col_listy = []\n",
    "for col in exp_summary_df.columns:\n",
    "    if '%' not in col:\n",
    "        col_listy.append(col)\n",
    "    else:\n",
    "        col_listy.append(col.replace('%', '_percentile'))\n",
    "print(col_listy)\n",
    "exp_summary_df.columns = col_listy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**And add some other dispersion metrics**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "exp_summary_df['cv'] = exp_summary_df['std']/exp_summary_df['mean']\n",
    "exp_summary_df['noise'] = exp_summary_df['std'].pow(2)/exp_summary_df['mean'].pow(2)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Finally, adding some of (what we think are) the cool new variables to consider**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "slopes = []\n",
    "pearsons = []\n",
    "spearmans = []\n",
    "for gene in exp_summary_df.index:\n",
    "    a, b, c, d, e = stats.linregress(new_exp_df.loc[gene], meta_df['Growth Rate (1/hr)'])\n",
    "    slopes.append(a)\n",
    "    pearsons.append(c)\n",
    "    rho, p = stats.spearmanr(new_exp_df.loc[gene], meta_df['Growth Rate (1/hr)'])\n",
    "    spearmans.append(rho)\n",
    "    \n",
    "exp_summary_df['lin_slope'] = slopes\n",
    "exp_summary_df['lin_r'] = pearsons\n",
    "exp_summary_df['spearmans_rho'] = spearmans"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Save some files\n",
    "\n",
    "This was the whole point of all the code above. Should have some straightforward data tables now"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3923, 103)\n",
      "(103, 27)\n",
      "(3923, 13)\n"
     ]
    }
   ],
   "source": [
    "print(new_exp_df.shape)\n",
    "print(meta_df.shape)\n",
    "print(exp_summary_df.shape)\n",
    "new_exp_df.to_csv('../Data/processed_data/processed_expression_ecoli.tsv', sep='\\t')\n",
    "meta_df.to_csv('../Data/processed_data/processed_metadata_ecoli.tsv', sep='\\t')\n",
    "exp_summary_df.to_csv('../Data/processed_data/processed_expression_summary_ecoli.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get a thinned down dataset as a robustness check\n",
    "\n",
    "Since conditions are correlated with one another, I'm constructing a more sparsely populated dataset where inter-condition correlations are minimized using a greedy algorithm. \n",
    "\n",
    "The algorithm works as follows:\n",
    "1. Find which two conditions are the most highly correlated across the entire all-to-all correlation matrix\n",
    "2. Randomly delete one of the two conditions in question\n",
    "3. Iterate to (1)\n",
    "4. Stop after reaching a pre-defined final dataset size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "random.seed(42)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fur__wt_fe</th>\n",
       "      <th>fur__delfur_fe2</th>\n",
       "      <th>omics__bw_ac</th>\n",
       "      <th>omics__bw_fum</th>\n",
       "      <th>omics__bw_glc</th>\n",
       "      <th>omics__bw_pyr</th>\n",
       "      <th>cra_crp__wt_ac</th>\n",
       "      <th>cra_crp__wt_fru</th>\n",
       "      <th>cra_crp__delcra_ac</th>\n",
       "      <th>cra_crp__delcra_fru</th>\n",
       "      <th>...</th>\n",
       "      <th>pal__tartr_ale26_f10</th>\n",
       "      <th>pal__tartr_ale26</th>\n",
       "      <th>pal__tartr_ale28_f10</th>\n",
       "      <th>pal__tartr_ale28</th>\n",
       "      <th>efeU__menFentC_ale39</th>\n",
       "      <th>efeU__menFentC_ale29</th>\n",
       "      <th>efeU__menFentC_ale30</th>\n",
       "      <th>efeU__menFentCubiC_ale36</th>\n",
       "      <th>efeU__menFentCubiC_ale37</th>\n",
       "      <th>efeU__menFentCubiC_ale38</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>fur__wt_fe</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.964640</td>\n",
       "      <td>0.923339</td>\n",
       "      <td>0.917715</td>\n",
       "      <td>0.962571</td>\n",
       "      <td>0.935009</td>\n",
       "      <td>0.901038</td>\n",
       "      <td>0.926609</td>\n",
       "      <td>0.904316</td>\n",
       "      <td>0.943159</td>\n",
       "      <td>...</td>\n",
       "      <td>0.866013</td>\n",
       "      <td>0.866432</td>\n",
       "      <td>0.858613</td>\n",
       "      <td>0.918781</td>\n",
       "      <td>0.873224</td>\n",
       "      <td>0.850732</td>\n",
       "      <td>0.928566</td>\n",
       "      <td>0.934770</td>\n",
       "      <td>0.939126</td>\n",
       "      <td>0.927869</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>fur__delfur_fe2</th>\n",
       "      <td>0.964640</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.913742</td>\n",
       "      <td>0.903783</td>\n",
       "      <td>0.952763</td>\n",
       "      <td>0.924470</td>\n",
       "      <td>0.898731</td>\n",
       "      <td>0.914061</td>\n",
       "      <td>0.924542</td>\n",
       "      <td>0.961016</td>\n",
       "      <td>...</td>\n",
       "      <td>0.854125</td>\n",
       "      <td>0.849878</td>\n",
       "      <td>0.858827</td>\n",
       "      <td>0.901197</td>\n",
       "      <td>0.840142</td>\n",
       "      <td>0.834696</td>\n",
       "      <td>0.896473</td>\n",
       "      <td>0.942488</td>\n",
       "      <td>0.946356</td>\n",
       "      <td>0.904614</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>omics__bw_ac</th>\n",
       "      <td>0.923339</td>\n",
       "      <td>0.913742</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.967975</td>\n",
       "      <td>0.942483</td>\n",
       "      <td>0.972111</td>\n",
       "      <td>0.943399</td>\n",
       "      <td>0.930222</td>\n",
       "      <td>0.931166</td>\n",
       "      <td>0.885828</td>\n",
       "      <td>...</td>\n",
       "      <td>0.907638</td>\n",
       "      <td>0.897741</td>\n",
       "      <td>0.903667</td>\n",
       "      <td>0.949034</td>\n",
       "      <td>0.839838</td>\n",
       "      <td>0.817468</td>\n",
       "      <td>0.878564</td>\n",
       "      <td>0.884267</td>\n",
       "      <td>0.884292</td>\n",
       "      <td>0.869539</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>omics__bw_fum</th>\n",
       "      <td>0.917715</td>\n",
       "      <td>0.903783</td>\n",
       "      <td>0.967975</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.956253</td>\n",
       "      <td>0.974362</td>\n",
       "      <td>0.902824</td>\n",
       "      <td>0.916864</td>\n",
       "      <td>0.892215</td>\n",
       "      <td>0.869514</td>\n",
       "      <td>...</td>\n",
       "      <td>0.867798</td>\n",
       "      <td>0.873566</td>\n",
       "      <td>0.850602</td>\n",
       "      <td>0.940838</td>\n",
       "      <td>0.842448</td>\n",
       "      <td>0.820857</td>\n",
       "      <td>0.897118</td>\n",
       "      <td>0.879182</td>\n",
       "      <td>0.874671</td>\n",
       "      <td>0.884392</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>omics__bw_glc</th>\n",
       "      <td>0.962571</td>\n",
       "      <td>0.952763</td>\n",
       "      <td>0.942483</td>\n",
       "      <td>0.956253</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.966576</td>\n",
       "      <td>0.885214</td>\n",
       "      <td>0.913576</td>\n",
       "      <td>0.897124</td>\n",
       "      <td>0.932456</td>\n",
       "      <td>...</td>\n",
       "      <td>0.856878</td>\n",
       "      <td>0.858610</td>\n",
       "      <td>0.856949</td>\n",
       "      <td>0.925330</td>\n",
       "      <td>0.867961</td>\n",
       "      <td>0.851744</td>\n",
       "      <td>0.924271</td>\n",
       "      <td>0.932579</td>\n",
       "      <td>0.932818</td>\n",
       "      <td>0.925066</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 103 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 fur__wt_fe  fur__delfur_fe2  omics__bw_ac  omics__bw_fum  \\\n",
       "fur__wt_fe         1.000000         0.964640      0.923339       0.917715   \n",
       "fur__delfur_fe2    0.964640         1.000000      0.913742       0.903783   \n",
       "omics__bw_ac       0.923339         0.913742      1.000000       0.967975   \n",
       "omics__bw_fum      0.917715         0.903783      0.967975       1.000000   \n",
       "omics__bw_glc      0.962571         0.952763      0.942483       0.956253   \n",
       "\n",
       "                 omics__bw_glc  omics__bw_pyr  cra_crp__wt_ac  \\\n",
       "fur__wt_fe            0.962571       0.935009        0.901038   \n",
       "fur__delfur_fe2       0.952763       0.924470        0.898731   \n",
       "omics__bw_ac          0.942483       0.972111        0.943399   \n",
       "omics__bw_fum         0.956253       0.974362        0.902824   \n",
       "omics__bw_glc         1.000000       0.966576        0.885214   \n",
       "\n",
       "                 cra_crp__wt_fru  cra_crp__delcra_ac  cra_crp__delcra_fru  \\\n",
       "fur__wt_fe              0.926609            0.904316             0.943159   \n",
       "fur__delfur_fe2         0.914061            0.924542             0.961016   \n",
       "omics__bw_ac            0.930222            0.931166             0.885828   \n",
       "omics__bw_fum           0.916864            0.892215             0.869514   \n",
       "omics__bw_glc           0.913576            0.897124             0.932456   \n",
       "\n",
       "                 ...  pal__tartr_ale26_f10  pal__tartr_ale26  \\\n",
       "fur__wt_fe       ...              0.866013          0.866432   \n",
       "fur__delfur_fe2  ...              0.854125          0.849878   \n",
       "omics__bw_ac     ...              0.907638          0.897741   \n",
       "omics__bw_fum    ...              0.867798          0.873566   \n",
       "omics__bw_glc    ...              0.856878          0.858610   \n",
       "\n",
       "                 pal__tartr_ale28_f10  pal__tartr_ale28  efeU__menFentC_ale39  \\\n",
       "fur__wt_fe                   0.858613          0.918781              0.873224   \n",
       "fur__delfur_fe2              0.858827          0.901197              0.840142   \n",
       "omics__bw_ac                 0.903667          0.949034              0.839838   \n",
       "omics__bw_fum                0.850602          0.940838              0.842448   \n",
       "omics__bw_glc                0.856949          0.925330              0.867961   \n",
       "\n",
       "                 efeU__menFentC_ale29  efeU__menFentC_ale30  \\\n",
       "fur__wt_fe                   0.850732              0.928566   \n",
       "fur__delfur_fe2              0.834696              0.896473   \n",
       "omics__bw_ac                 0.817468              0.878564   \n",
       "omics__bw_fum                0.820857              0.897118   \n",
       "omics__bw_glc                0.851744              0.924271   \n",
       "\n",
       "                 efeU__menFentCubiC_ale36  efeU__menFentCubiC_ale37  \\\n",
       "fur__wt_fe                       0.934770                  0.939126   \n",
       "fur__delfur_fe2                  0.942488                  0.946356   \n",
       "omics__bw_ac                     0.884267                  0.884292   \n",
       "omics__bw_fum                    0.879182                  0.874671   \n",
       "omics__bw_glc                    0.932579                  0.932818   \n",
       "\n",
       "                 efeU__menFentCubiC_ale38  \n",
       "fur__wt_fe                       0.927869  \n",
       "fur__delfur_fe2                  0.904614  \n",
       "omics__bw_ac                     0.869539  \n",
       "omics__bw_fum                    0.884392  \n",
       "omics__bw_glc                    0.925066  \n",
       "\n",
       "[5 rows x 103 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cond_corr_mat = new_exp_df.corr(method='spearman')\n",
    "cond_corr_mat.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**The `final_data_size` is of course completely arbitrary. But the point is to get rid of some correlated data so it does the trick.**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "final_data_size = 30\n",
    "n_to_prune = cond_corr_mat.shape[1] - final_data_size"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.fill_diagonal(cond_corr_mat.values, np.nan)\n",
    "to_prune = []\n",
    "for i in range(n_to_prune):\n",
    "    tempy = cond_corr_mat.loc[[i for i in cond_corr_mat.columns if i not in to_prune]][[i for i in cond_corr_mat.columns if i not in to_prune]].max()\n",
    "    to_prune.append(random.choice([tempy.sort_values(ascending=False).index[0],\\\n",
    "                                   tempy.sort_values(ascending=False).index[1]]))\n",
    "np.fill_diagonal(cond_corr_mat.values, 1.)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Remove relevant rows from two of the dataframes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "to_keep = [i for i in cond_corr_mat.columns if i not in to_prune]\n",
    "new_exp_df = new_exp_df[to_keep]\n",
    "meta_df = meta_df[meta_df['Simple_sample_id'].isin(to_keep)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**And regenerate the summary dataframe to only use this thinned-down set of samples**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['count', 'mean', 'std', 'min', '25_percentile', '50_percentile', '75_percentile', 'max']\n"
     ]
    }
   ],
   "source": [
    "exp_summary_df = new_exp_df.apply(pd.DataFrame.describe, axis=1)\n",
    "col_listy = []\n",
    "for col in exp_summary_df.columns:\n",
    "    if '%' not in col:\n",
    "        col_listy.append(col)\n",
    "    else:\n",
    "        col_listy.append(col.replace('%', '_percentile'))\n",
    "print(col_listy)\n",
    "exp_summary_df.columns = col_listy\n",
    "\n",
    "exp_summary_df['cv'] = exp_summary_df['std']/exp_summary_df['mean']\n",
    "exp_summary_df['noise'] = exp_summary_df['std'].pow(2)/exp_summary_df['mean'].pow(2)\n",
    "\n",
    "slopes = []\n",
    "pearsons = []\n",
    "spearmans = []\n",
    "for gene in exp_summary_df.index:\n",
    "    a, b, c, d, e = stats.linregress(new_exp_df.loc[gene], meta_df['Growth Rate (1/hr)'])\n",
    "    slopes.append(a)\n",
    "    pearsons.append(c)\n",
    "    rho, p = stats.spearmanr(new_exp_df.loc[gene], meta_df['Growth Rate (1/hr)'])\n",
    "    spearmans.append(rho)\n",
    "    \n",
    "exp_summary_df['lin_slope'] = slopes\n",
    "exp_summary_df['lin_r'] = pearsons\n",
    "exp_summary_df['spearmans_rho'] = spearmans"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(3923, 30)\n",
      "(30, 27)\n",
      "(3923, 13)\n"
     ]
    }
   ],
   "source": [
    "print(new_exp_df.shape)\n",
    "print(meta_df.shape)\n",
    "print(exp_summary_df.shape)\n",
    "new_exp_df.to_csv('../Data/processed_data/processed_expression_ecoli_SPARSE.tsv', sep='\\t')\n",
    "meta_df.to_csv('../Data/processed_data/processed_metadata_ecoli_SPARSE.tsv', sep='\\t')\n",
    "exp_summary_df.to_csv('../Data/processed_data/processed_expression_summary_ecoli_SPARSE.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Scratch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  },
  "toc": {
   "colors": {
    "hover_highlight": "#DAA520",
    "navigate_num": "#000000",
    "navigate_text": "#333333",
    "running_highlight": "#FF0000",
    "selected_highlight": "#FFD700",
    "sidebar_border": "#EEEEEE",
    "wrapper_background": "#FFFFFF"
   },
   "moveMenuLeft": true,
   "nav_menu": {
    "height": "192px",
    "width": "252px"
   },
   "navigate_menu": true,
   "number_sections": true,
   "sideBar": true,
   "threshold": 4,
   "toc_cell": false,
   "toc_section_display": "block",
   "toc_window_display": false,
   "widenNotebook": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
