---
title: "R Notebook"
output: html_notebook
---


```{r}
library("AnaCoDa")
```


```{r}
#genome <- initializeGenomeObject(file = "../../Data/ecoli_info/NC_000913.3.CLEAN.CDS.fasta")
genome <- initializeGenomeObject(file = "../../Data/scer_info/SCER.CLEAN.CDS.fasta")

```


```{r, echo=FALSE, message=FALSE, warning=FALSE, include=FALSE}
parameter <- initializeParameterObject(genome = genome, sphi = 1, num.mixtures = 1, gene.assignment = rep(1, length(genome)))
model <- initializeModelObject(parameter = parameter, model = "ROC")
mcmc <- initializeMCMCObject(samples = 5000, thinning = 10, adaptive.width=50)
runMCMC(mcmc = mcmc, genome = genome, model = model, ncores = 1)
```


```{r}
phi_mat <- getExpressionEstimates(parameter = parameter, 
                                  gene.index = 1:length(genome),
                                  samples = 5000, genome = genome)
head(phi_mat)
```

```{r}
#write.csv(phi_mat,"../../Data/ecoli_info/roc_semppr_ecoli.csv", row.names = FALSE, quote=FALSE)
write.csv(phi_mat,"../../Data/scer_info/roc_semppr_scer.csv", row.names = FALSE, quote=FALSE)
```

